From 1348588f824ec142f59681a059b56b7208b57b82 Mon Sep 17 00:00:00 2001
From: Naveen M K <naveen521kk@gmail.com>
Date: Sun, 5 Jun 2022 20:28:53 +0530
Subject: [PATCH 053/N] tests: Fix some failing tests

- In `test_sysconfig`, ignore `test_sysconfig.TestSysConfig.test_user_similar` test failure.
- Copy `get_platform()` from from distutils.utils to test_importlib/test_windows.py.
- some basic fixes for test_getpath
- test_sysconfig.py: fix tests related to mingw
- Use bin instead of Scripts in venv
---
 Lib/test/support/__init__.py            |  2 +-
 Lib/test/test_getpath.py                |  3 +++
 Lib/test/test_importlib/test_windows.py | 23 +++++++++++++++++++++++
 Lib/test/test_sysconfig.py              |  6 +++++-
 4 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/Lib/test/support/__init__.py b/Lib/test/support/__init__.py
index 780f789..e2d8744 100644
--- a/Lib/test/support/__init__.py
+++ b/Lib/test/support/__init__.py
@@ -2554,7 +2554,7 @@ def setup_venv_with_pip_setuptools(venv_dir):
 
         # Get the Python executable of the venv
         python_exe = os.path.basename(sys.executable)
-        if sys.platform == 'win32':
+        if sys.platform == 'win32' and not sysconfig.get_platform().startswith('mingw'):
             python = os.path.join(venv, 'Scripts', python_exe)
         else:
             python = os.path.join(venv, 'bin', python_exe)
diff --git a/Lib/test/test_getpath.py b/Lib/test/test_getpath.py
index 2907abe..d12cb5c 100644
--- a/Lib/test/test_getpath.py
+++ b/Lib/test/test_getpath.py
@@ -972,6 +972,7 @@ DEFAULT_NAMESPACE = dict(
     ENV_PYTHONHOME="",
     ENV_PYTHONEXECUTABLE="",
     ENV___PYVENV_LAUNCHER__="",
+    ENV_MSYSTEM="",
     argv0="",
     py_setpath="",
     real_executable="",
@@ -1011,6 +1012,7 @@ class MockNTNamespace(dict):
         self.update(DEFAULT_NAMESPACE)
         self["config"] = DEFAULT_CONFIG.copy()
         self["os_name"] = "nt"
+        self["is_mingw"] = 0
         self["PLATLIBDIR"] = "DLLs"
         self["PYWINVER"] = "9.8-XY"
         self["VPATH"] = r"..\.."
@@ -1187,6 +1189,7 @@ class MockPosixNamespace(dict):
         self.update(DEFAULT_NAMESPACE)
         self["config"] = DEFAULT_CONFIG.copy()
         self["os_name"] = "posix"
+        self["is_mingw"] = 0
         self["PLATLIBDIR"] = "lib"
         self["WITH_NEXT_FRAMEWORK"] = 0
         super().__init__(*a, **kw)
diff --git a/Lib/test/test_importlib/test_windows.py b/Lib/test/test_importlib/test_windows.py
index bef4fb4..3098111 100644
--- a/Lib/test/test_importlib/test_windows.py
+++ b/Lib/test/test_importlib/test_windows.py
@@ -24,6 +24,29 @@ def get_platform():
             'x64' : 'win-amd64',
             'arm' : 'win-arm32',
         }
+    if os.name == 'nt':
+        if sys._is_mingw:
+            platform = "mingw"
+            if "amd64" in sys.version.lower():
+                platform += "_x86_64"
+            elif "arm64" in sys.version.lower():
+                platform += "_aarch64"
+            elif "arm" in sys.version.lower():
+                platform += "_armv7"
+            else:
+                platform += "_i686"
+
+            if "ucrt" in sys.version.lower():
+                platform += "_ucrt"
+            else:
+                platform += "_msvcrt"
+
+            if "clang" in sys.version.lower():
+                platform += "_llvm"
+            else:
+                platform += "_gnu"
+
+            return platform
     if ('VSCMD_ARG_TGT_ARCH' in os.environ and
         os.environ['VSCMD_ARG_TGT_ARCH'] in TARGET_TO_PLAT):
         return TARGET_TO_PLAT[os.environ['VSCMD_ARG_TGT_ARCH']]
diff --git a/Lib/test/test_sysconfig.py b/Lib/test/test_sysconfig.py
index 5a5f48a..844085f 100644
--- a/Lib/test/test_sysconfig.py
+++ b/Lib/test/test_sysconfig.py
@@ -210,7 +210,7 @@ class TestSysConfig(unittest.TestCase, VirtualEnvironmentMixin):
         self.assertEqual(libpath, sysconfig.get_path('purelib', scheme='nt_venv', vars=vars))
 
     def test_venv_scheme(self):
-        if sys.platform == 'win32':
+        if not sys._is_mingw and sys.platform == 'win32':
             self.assertEqual(
                 sysconfig.get_path('scripts', scheme='venv'),
                 sysconfig.get_path('scripts', scheme='nt_venv')
@@ -432,6 +432,10 @@ class TestSysConfig(unittest.TestCase, VirtualEnvironmentMixin):
             if HAS_USER_BASE:
                 user_path = get_path(name, 'posix_user')
                 expected = os.path.normpath(global_path.replace(base, user, 1))
+                if os.name == 'nt' and sys._is_mingw:
+                    expected = expected.replace(
+                        f'python{sysconfig.get_python_version()}',
+                        f'python{sysconfig.get_python_version()}-{get_platform()}')
                 # bpo-44860: platlib of posix_user doesn't use sys.platlibdir,
                 # whereas posix_prefix does.
                 if name == 'platlib':
